SemanticUI/Modules/Dropdown.elm

```diff
@@ -0,0 +1,456 @@
+            ((classList
+                [ ( "menu", True )
+                , ( "active", True )
+                , ( "visible", ctx.drawerIsVisible )
+                , ( "hiden", not ctx.drawerIsVisible )
+                , ( "animating", ctx.drawerIsTransitioning )
+                , ( "transition", True )
+                , ( "slide", ctx.drawerIsTransitioning )
+                , ( "down", ctx.drawerIsTransitioning )
+                , ( "in", state == Opening )
+                , ( "out", state == Closing )
+                ]
+                :: (if ctx.drawerIsTransitioning then
+                        [ on "animationend" (Decode.succeed (cfg.onToggle <| ctx.drawerToggledFinal)) ]
+
+                    else
+                        []
+                   )
+             )
+                ++ attrs
+            )
+            children
+            ctx.drawerIsOpen
+            ctx.dropdownCfg
+
+
+{-| Create an item that goes in the drawer.
+
+It takes the following:
+
+  - `Config`
+  - `State`
+  - Node type as a render function e.g. `Html.div`
+  - Attributes of node
+  - Children of node
+
+It adds the "item" class to the node.
+
+-}
+item : Render msg -> List (Attribute msg) -> List (Html msg) -> Html msg
+item render attrs children =
+    render (class "item" :: attrs) children
+
+
+type alias Context msg =
+    { drawerIsOpen : Bool
+    , drawerIsTransitioning : Bool
+    , drawerIsVisible : Bool
+    , drawerToggledFinal : State
+    , dropdownCfg : Dropdown.Config msg
+    }
+
+
+toContext : Config r msg -> State -> Context msg
+toContext cfg state =
+    let
+        drawerIsOpen =
+            state == Opening || state == Opened
+
+        drawerIsTransitioning =
+            state == Opening || state == Closing
+
+        drawerIsVisible =
+            drawerIsOpen || drawerIsTransitioning
+
+        drawerToggledFinal =
+            if drawerIsTransitioning then
+                if state == Opening then
+                    Opened
+
+                else
+                    Closed
+
+            else
+                state
+
+        dropdownCfg =
+            { identifier = cfg.identifier
+            , toggleEvent =
+                case cfg.toggleEvent of
+                    OnClick ->
+                        Dropdown.OnClick
+
+                    OnHover ->
+                        Dropdown.OnHover
+
+                    OnFocus ->
+                        Dropdown.OnFocus
+            , drawerVisibleAttribute = classList []
+            , callback =
+                cfg.onToggle
+                    << (\b ->
+                            if b then
+                                if drawerIsOpen then
+                                    state
+
+                                else
+                                    Opening
+
+                            else if not drawerIsOpen then
+                                state
+
+                            else
+                                Closing
```

Comment:
I tend to like listing out the truth table for cases like this where every combination needs to be analyzed. Ollie, would you mind if I changed it to read like this?

```elm
case ( toggleOpen, isOpen ) of
    ( True, False ) ->
        Opening

    ( False, True ) ->
        Closing

    _ ->
        state

```